### PR TITLE
[HttpFoundation] JSONP callback validation

### PR DESCRIPTION
**Pull request by ro0NL**
| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | yes |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #17923 |
| License | MIT |
| Doc PR | reference to the documentation PR, if any |

Maybe this is too small for a new dep, but at least it's stable. Symfony itself will make no assumption on validation by default, ie. things should keep working as usual.

### CODE REVIEW

**Comment by javiereguiluz** [comment]
For now I'm 👎  because the replaced feature is minuscule and introducing a new dependency is not worth it (specially now, that Symfony is furiously removing dependencies ... even between its components).

**Comment by theofidry** [comment]
To be fair @javiereguiluz the dependency in question is fairly small as well, its scope is the scope of the feature.

**Comment by ro0NL** [comment]
Understand.. however if im correct the philosophy is to not reinvent the wheel.

Problem is we cant workaround symfony's out-of-the-box validation.

**Comment by theofidry** [comment]
@ro0NL you can workaround it by overriding it, it's a public method. You're PR is not changing anything in this regard.

**Comment by wouterj** [comment]
please note that this is introducing an **optional dependency**.

**Comment by ro0NL** [comment]
@theofidry  dair enough :) but still is cumbersome imo.

**Comment by ro0NL** [comment]
We could also just copy the "right" regex.. i'd prefer the dependency though.

**Comment by fabpot** [comment]
IIUC, if you don't install the dependency, the callback is not validated anymore?

Anyway, I don't see why we should depend on an external library for that. Is there any issue with our code? If yes, let's fix it. But as that's not something that changes over time, I don't see the advantage of depending on something else.

**Comment by ro0NL** [comment]
> if you don't install the dependency, the callback is not validated anymore?

Correct.. I dont mind forcing validation either, but if so, it should be correct. So basically there are 4 viable options;
- make it a hard dependency
- copy/fix the regex

Im not sure we should we should keep patching our regex whenever it shows being incorrect, ie. a stable/tested version is available => https://github.com/willdurand/JsonpCallbackValidator.

**Comment by fabpot** [comment]
Let's fix our regex.

**Comment by ro0NL** [comment]
Shall we just take https://github.com/willdurand/JsonpCallbackValidator/blob/master/src/JsonpCallbackValidator.php#L11 then? And forget about reserved keywords.. or take those as well?

**Comment by fabpot** [comment]
You can take everything (and don't forget to include all the author/license information with it).

**Comment by ro0NL** [comment]
:+1: ill have a look tonight.

**Comment by ro0NL** [comment]
@fabpot ready. I kept the unicode support as well.. so basically we only added support for bracket notation.. (taken from willdurand/JsonpCallbackValidator).

https://regex101.com/r/bSdU1i/2

File: src/Symfony/Component/HttpFoundation/JsonResponse.php [hunk 1]
```diff
@@ -89,11 +89,16 @@ public static function fromJsonString($data = null, $status = 200, $headers = ar
     public function setCallback($callback = null)
     {
         if (null !== $callback) {
-            // taken from http://www.geekality.net/2011/08/03/valid-javascript-identifier/
-            $pattern = '/^[$_\p{L}][$_\p{L}\p{Mn}\p{Mc}\p{Nd}\p{Pc}\x{200C}\x{200D}]*+$/u';
```
Comment (ro0NL) [inline src/Symfony/Component/HttpFoundation/JsonResponse.php]: Removed the possessive `+` as this wont affect number of steps to match due `^$`.

File: src/Symfony/Component/HttpFoundation/JsonResponse.php [hunk 2]
```diff
@@ -89,11 +89,16 @@ public static function fromJsonString($data = null, $status = 200, $headers = ar
...
+            // taken from http://www.geekality.net/2011/08/03/valid-javascript-identifier/ and https://github.com/willdurand/JsonpCallbackValidator
+            $pattern = '/^[$_\p{L}][$_\p{L}\p{Mn}\p{Mc}\p{Nd}\p{Pc}\x{200C}\x{200D}]*(?:\[(?:"(?:\\\.|[^"\\\])*"|\'(?:\\\.|[^\'\\\])*\'|\d+)\])*?$/u';
```
Comment (fabpot) [inline src/Symfony/Component/HttpFoundation/JsonResponse.php]: You need to add the copyright and the license

Comment (ro0NL) [inline src/Symfony/Component/HttpFoundation/JsonResponse.php]: Can you have a look.. not sure i did it correct.

**Comment by fabpot** [comment]
👍

**Comment by dunglas** [comment]
👍

**Comment by fabpot** [comment]
Thank you @ro0NL.

**Pull request closed without merging**